pym/bob/cmds/build/share.py

```diff
@@ -0,0 +1,301 @@
+                try:
+                    with OpenLocked(fn, "x", True) as f:
+                        json.dump({"pkgs" : {asHexStr(buildId) : size}}, f)
+                        return size
+                except FileExistsError:
+                    # Almost impossible case: lost creation race -> update
+                    with OpenLocked(fn, "r+", True) as f:
+                        return update(f)
+        except OSError as e:
+            raise BuildError("Error updating shared repo: "+str(e))
+
+    def canInstall(self, buildId):
+        path = self.__buildPath(buildId)
+        return not os.path.isdir(path)
+
+    def useSharedPackage(self, workspace, buildId):
+        path = self.__buildPath(buildId)
+        workspace = os.path.abspath(workspace)
+        try:
+            with OpenLocked(os.path.join(self.__path, "repo.json"), "r", False):
+                pkgMetaFile = os.path.join(path, "pkg.json")
+                with OpenLocked(pkgMetaFile, "r+", True) as f:
+                    if not os.path.isdir(path):
+                        return None, None # concurrent gc wiped the package :(
+                    meta = json.load(f)
+                    sharedHash = bytes.fromhex(meta.get("hash", ""))
+                    if len(sharedHash) != 20:
+                        raise BuildError("Invalid shared result hash in " + path)
+
+                    # Make sure our workspace is recorded in the metainfo.
+                    users = meta.setdefault("users", [])
+                    if workspace not in users:
+                        users.append(workspace)
+                        f.seek(0)
+                        f.truncate()
+                        json.dump(meta, f)
+                    else:
+                        os.utime(pkgMetaFile)
+        except FileNotFoundError:
+            return None, None
+        except OSError as e:
+            raise BuildError("Could not read shared result meta info: " + str(e))
+        except (json.JSONDecodeError, ValueError) as e:
+            raise BuildError("Corrupt meta info in {}: {}".format(path, str(e)))
+
+        return path, sharedHash
+
+    def installSharedPackage(self, workspace, buildId, sharedHash, mayMove):
+        # Quick check: was somebody faster?
+        sharedPath = self.__buildPath(buildId)
+        if os.path.isdir(sharedPath):
+            return sharedPath, False
+
+        # Prepare everyting in temporary directory next to the shared packages
+        # to atomically "install" the whole package with a single move. Can
+        # still lose the race at the final rename!
+        repoSize = 0
+        try:
+            os.makedirs(os.path.dirname(sharedPath), exist_ok=True)
+            with tempfile.TemporaryDirectory(dir=self.__path) as tmpDir:
+                tmpSharedPath = os.path.join(tmpDir, "pkg")
+                os.mkdir(tmpSharedPath)
+                shutil.copyfile(os.path.join(workspace, "..", "audit.json.gz"),
```

Comment:
If audit is disabled, the copy operation will fail. I guess we should add a checker if audit is active or not?

Reply:
I'm more inclined to prohibit installation of packages that were built without audit trail. Right now such packages are not uploaded either. Would that be a problem?

Reply:
Ok, that makes sense of course. I just stumbled across it while testing with disabled audit trail.